target_chains/ethereum/contracts/forge-test/VerificationExperiments.t.sol

```diff
@@ -0,0 +1,600 @@
+
+        (
+            PythInternalStructs.PriceInfo memory info,
+            bytes32 priceId
+        ) = parseSingleAttestationFromBatch(data, 0, data.length);
+        uint64 latestPublishTime = latestPriceInfoPublishTime(priceId);
+
+        if (info.publishTime > latestPublishTime) {
+            setLatestPriceInfo(priceId, info);
+            emit PriceFeedUpdate(
+                priceId,
+                info.publishTime,
+                info.price,
+                info.conf
+            );
+        }
+    }
+
+    // Update a single price feed via a threshold-signed price update.
+    // data is expected to be a serialized PriceAttestation.
+    function updatePriceFeedsThreshold(
+        bytes memory signature,
+        bytes memory data
+    ) public payable {
+        bytes32 hash = keccak256(data);
+        if (!verifySignature(hash, signature, thresholdPublicKey))
+            revert PythErrors.InvalidArgument();
+
+        (
+            PythInternalStructs.PriceInfo memory info,
+            bytes32 priceId
+        ) = parseSingleAttestationFromBatch(data, 0, data.length);
+        uint64 latestPublishTime = latestPriceInfoPublishTime(priceId);
+
+        if (info.publishTime > latestPublishTime) {
+            setLatestPriceInfo(priceId, info);
+            emit PriceFeedUpdate(
+                priceId,
+                info.publishTime,
+                info.price,
+                info.conf
+            );
+        }
+    }
+
+    // Update a single price feed via a "native" price update (i.e., using the default ethereum tx signature for authentication).
+    // data is expected to be a serialized PriceAttestation.
+    // This function represents the lower bound on how much gas we can use.
+    function updatePriceFeedsNative(bytes memory data) public payable {
+        // TODO: this function should have a check on the sender.
+        // I'm assuming that check is not very expensive here.
+
+        (
+            PythInternalStructs.PriceInfo memory info,
+            bytes32 priceId
+        ) = parseSingleAttestationFromBatch(data, 0, data.length);
+        uint64 latestPublishTime = latestPriceInfoPublishTime(priceId);
+
+        if (info.publishTime > latestPublishTime) {
+            setLatestPriceInfo(priceId, info);
+            emit PriceFeedUpdate(
+                priceId,
+                info.publishTime,
+                info.price,
+                info.conf
+            );
+        }
+    }
+
+    // Verify that signature is a valid ECDSA signature of messageHash by signer.
+    function verifySignature(
+        bytes32 messageHash,
+        bytes memory signature,
+        address signer
+    ) public pure returns (bool) {
+        bytes32 r;
+        bytes32 s;
+        uint8 v;
+        assembly {
+            r := mload(add(signature, 32))
+            s := mload(add(signature, 64))
+            v := byte(0, mload(add(signature, 96)))
+        }
```

Comment:
Nitpik: I think we can do it easily without this using the byteslib that we have. I prefer not to have assembly code here.